dkg/sync/client.go

```diff
@@ -53,7 +53,7 @@ func NewClient(tcpNode host.Host, peer peer.ID, hashSig []byte, version version.
 // supports reestablishing on relay circuit recycling, and supports soft shutdown.
 type Client struct {
 	// Mutable state
-	mu        sync.Mutex
+	mu        sync.RWMutex
```

Comment:
why do we need a read-write mutex here?

Reply:
Because in many places we need just read. Using `RWMutex` optimizes performance (less locks and waits) and is widely adopted in golang applications. However, I found that Charon often uses normal Mutex, and it is unclear why. I would wish to eventually replace everything to RWMutex where appropriate.

Reply:
@pinebit Absolutely agree that `RWMutex`es are more optimized. But wondering if we necessarily require this. In charon, we have tried to refrain from doing "premature optimizations" wherever possible, hence you would see more mutexes used than rw-mutexes.

By what magnitude do you believe using `RWMutex` can optimize the performance here?

Else, we can discuss this async and create a ticket to refactor mutexes in the codebase to use `RWMutex` in a separate PR

Reply:
This is more about training muscle memory and develop good habits to use the appropriate API. Not necessary to optimize the performance in this particular case. Also, this improves readability - you show an intention. For example if I see RLock(), I can quickly realize this function should not modify the state and then I will think twice when introducing a change.

Reply:
> For example if I see RLock(), I can quickly realize this function should not modify the state and then I will think twice when introducing a change.
Yes precisely, this removes `congnitive load`!